File: source/ts/src/calendar/eclipsesTable.ts
// Engine-free reader: astronomy imports here would drag the engine into a client bundle.

import type {
  EclipsesFile,
  EclipsesTableLanguage,
  EclipseTableDay,
  EclipseTableEntry,
  EclipseTableEntryRaw,
} from './eclipsesTableTypes';

export type {
  EclipsesTableLanguage,
  EclipseTableKind,
  EclipseTableSubtype,
  EclipseSutak,
  EclipseTableEntryRaw,
  EclipseTableEntry,
  EclipseTableDay,
  EclipseTableMeta,
  RawEclipseTableDay,
  EclipsesFile,
} from './eclipsesTableTypes';

export function readEclipsesYearRange(
  source: EclipsesFile,
): { start: number; end: number } {
  return { start: source._meta.startYear, end: source._meta.endYear };
}

function flatten(
  raw: EclipseTableEntryRaw,
  lang: EclipsesTableLanguage,
): EclipseTableEntry {
  const name = raw.name[lang] ?? Object.values(raw.name)[0] ?? '';
  const out: EclipseTableEntry = {
    name,
    kind: raw.kind,
    subtype: raw.subtype,
    start: raw.start,
    peak: raw.peak,
    end: raw.end,
    obscuration: raw.obscuration,
    magnitude: raw.magnitude,
    visibleFromLocation: raw.visibleFromLocation,
    visibleAtPeak: raw.visibleAtPeak,
  };
  if (raw.sutak) out.sutak = raw.sutak;
  if (raw.description) {
    out.description = raw.description[lang] ?? Object.values(raw.description)[0] ?? '';
  }
  return out;
}

/** Eclipse days for a year; `null` out of range, `[]` in range with none. */
export function readEclipsesForYear(
  source: EclipsesFile,
  year: number,
  lang: EclipsesTableLanguage = 'en',
): EclipseTableDay[] | null {
  const days = source.years[String(year)];
  if (!days) return null;
  return days.map(d => ({
    date: d.date,
    eclipses: d.eclipses.map(e => flatten(e, lang)),
  }));
}

/** Eclipses on one date, matched on the peak's local date. */
export function readEclipsesForDate(
  source: EclipsesFile,
  date: string | Date,
  lang: EclipsesTableLanguage = 'en',
): EclipseTableEntry[] {
  const key = typeof date === 'string'
    ? date
    : toDateKey(date, source._meta.timezoneOffsetMinutes);
  const yearKey = key.slice(0, 4);
  const days = source.years[yearKey];
  if (!days) return [];
  const day = days.find(d => d.date === key);
  return day ? day.eclipses.map(e => flatten(e, lang)) : [];
}

function toDateKey(d: Date, offsetMinutes: number): string {
  const shifted = new Date(d.getTime() + offsetMinutes * 60_000);
  const y = shifted.getUTCFullYear();
  const m = String(shifted.getUTCMonth() + 1).padStart(2, '0');
  const day = String(shifted.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

/** @deprecated Renamed to {@link readEclipsesForYear} in v5. */
export const getEclipsesForYear = readEclipsesForYear;

/** @deprecated Renamed to {@link readEclipsesForDate} in v5. */
export const getEclipsesForDate = readEclipsesForDate;

/** @deprecated Renamed to {@link readEclipsesYearRange} in v5. */
export const getEclipsesYearRange = readEclipsesYearRange;

File: source/go/internal/calendar/eclipsestable.go
package calendar

func ReadEclipsesYearRange(source EclipsesFile) MuhurtaYearRangeLike {
	return MuhurtaYearRangeLike{Start: source.Meta.StartYear, End: source.Meta.EndYear}
}

func flattenEclipse(raw EclipseTableEntryRaw, lang FestivalsTableLanguage) EclipseTableEntry {
	out := EclipseTableEntry{
		Name:                raw.Name.Pick(lang),
		Kind:                raw.Kind,
		Subtype:             raw.Subtype,
		Start:               raw.Start,
		Peak:                raw.Peak,
		End:                 raw.End,
		Obscuration:         raw.Obscuration,
		Magnitude:           raw.Magnitude,
		VisibleFromLocation: raw.VisibleFromLocation,
		VisibleAtPeak:       raw.VisibleAtPeak,
	}
	if raw.Sutak != nil {
		s := *raw.Sutak
		out.Sutak = &s
	}
	if raw.Description != nil {
		out.Description = raw.Description.Pick(lang)
		out.HasDescription = true
	}
	return out
}

func ReadEclipsesForYear(
	source EclipsesFile, year int, lang FestivalsTableLanguage,
) ([]EclipseTableDay, bool) {
	days, ok := source.Years[itoa(year)]
	if !ok || days == nil {
		return nil, false
	}
	out := make([]EclipseTableDay, 0, len(days))
	for _, d := range days {
		entries := make([]EclipseTableEntry, 0, len(d.Eclipses))
		for _, e := range d.Eclipses {
			entries = append(entries, flattenEclipse(e, lang))
		}
		out = append(out, EclipseTableDay{Date: d.Date, Eclipses: entries})
	}
	return out, true
}

func ReadEclipsesForDateKey(
	source EclipsesFile, key string, lang FestivalsTableLanguage,
) []EclipseTableEntry {
	if len(key) < 4 {
		return []EclipseTableEntry{}
	}
	days, ok := source.Years[key[:4]]
	if !ok {
		return []EclipseTableEntry{}
	}
	for _, d := range days {
		if d.Date == key {
			out := make([]EclipseTableEntry, 0, len(d.Eclipses))
			for _, e := range d.Eclipses {
				out = append(out, flattenEclipse(e, lang))
			}
			return out
		}
	}
	return []EclipseTableEntry{}
}

func ReadEclipsesForDate(
	source EclipsesFile, dateMs int64, lang FestivalsTableLanguage,
) []EclipseTableEntry {
	return ReadEclipsesForDateKey(source,
		tableDateKey(dateMs, source.Meta.TimezoneOffsetMinutes), lang)
}
